### PR TITLE
Call predefined callable (invokable) with custom arguments per invokation

### PR DESCRIPTION
**Pull request by hannesvdvreken**
Helps out with thephpleague/route#12

This is a WIP to show how it could be done. Needs testing, explanation and documentation.

### CODE REVIEW

**Comment by hannesvdvreken** [comment]
I'll redo this with more atomic commits but wanted it pushed fast so I could shut down to enjoy the rest of the weekend.

**Comment by hannesvdvreken** [comment]
The gist I used to test it: https://gist.github.com/hannesvdvreken/8b83f00ebf0c69cb4555

**Comment by philipobenito** [comment]
Any chance of a test case just to be safe then I'll merge? Then on to v2, massive tidy up :-)

**Comment by hannesvdvreken** [comment]
I was still going to add some tests. Will still keep that to do on my list so it's safer to refactor.

**Comment by philipobenito** [comment]
Thanks :-)

**Comment by sagikazarmark** [comment]
This PR screwed something up, there is some problem with the array_map in CallableDefinition. Also, I vote for \Reflection* classes NOT being aliased to the current namespace, they are noise.

**Comment by sagikazarmark** [comment]
Errors I am getting:

* Warning: array_map(): An error occurred while invoking the map callback
* Fatal error: Uncaught exception 'RuntimeException' with message 'Cannot resolve argument [name], it should be provided within an array of arguments passed to [Fuel\Dependency\Container::call], have a default value or be type hinted'

**Comment by hannesvdvreken** [comment]
@sagikazarmark No need to argue about import statements.

I assume you mean [this](https://github.com/thephpleague/container/blob/806ba15e2b7056b387d1d579e57f4496875d038a/src/Definition/CallableDefinition.php#L75) `array_map`?

File: src/ContainerInterface.php [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 namespace League\Container;
 
-interface ContainerInterface
+use ArrayAccess;
+
+interface ContainerInterface extends ArrayAccess
```
Comment (sagikazarmark) [inline src/ContainerInterface.php]: Why did array access become part of the interface?

**Comment by sagikazarmark** [comment]
Yes, I do

**Comment by sagikazarmark** [comment]
@hannesvdvreken I am trying to be consistent, seems like the rest of the library does not use unnecessary imports.

**Comment by philipobenito** [comment]
@sagikazarmark import everything seems to be a league standard and are being picked up as and when seen.

As for the error you've shown, that's an exception you are throwing, not from the container so I wouldn't know the issue without looking through your lib.

**Comment by sagikazarmark** [comment]
The exception is NOT thrown by my code (but the Container, see the exception message), but we sorted it out with @hannesvdvreken on IRC. There is a BC break, @hannesvdvreken will push a fix when he has time. (He is going to the gym now :stuck_out_tongue_winking_eye:)

**Comment by sagikazarmark** [comment]
Ah, sorry, correction. @hannesvdvreken is not going to the gym NOW. :stuck_out_tongue_winking_eye:

**Comment by hannesvdvreken** [comment]
@philipobenito thanks for backing me on the import stuff ;-)

So, @sagikazarmark and I have been chatting on IRC about this.
It appears we broke backwards compatibility here.

This used to work, but now it isn't anymore.

```php
$container->add('notifier.sms', function ($token, $ssl) {
    return new \Twilio\SMS($token, $ssl);
});

$container->get('notifier.sms', ['abcd', true]);
```

I didn't notice when registering such a callback it actually created a CallableDefinition. And that's where things changed (and broke BC).

I'm thinking about writing some tests for this first and then solve it by creating a different kind of Definition that doesn't have `$this->callables` and just passes on the `$vars` to the callable without reflection. This way BC should be restored.

**Comment by philipobenito** [comment]
@sagikazarmark sorry didn't recognise the message. At work so only scanned. I'll remove the tag and release for now until fixed.

**Pull request merged**